helper/serialization.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helper
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/hcl/hcl/printer"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	tfschema "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// InstanceStateToHCL converts a terraform.InstanceState into the HCL that represents the maximally hydrated form.  The generated
+// HCL will not include any references to other resources, since that information cannot be reconstructed without a full view
+// of all resources present..
+func InstanceStateToHCL(state *terraform.InstanceState, info *terraform.InstanceInfo, provider *tfschema.Provider) (string, error) {
+	providerSchema := provider.ResourcesMap[info.Type].Schema
+	str, err := resourceOrSubresourceHCL(providerSchema, state.Attributes)
+	if err != nil {
+		return "", err
+	}
+	str = fmt.Sprintf("resource %q %q {\n%s\n}\n", info.Type, info.Id, str)
+	hBytes, err := printer.Format([]byte(str))
+	if err != nil {
+		return "", fmt.Errorf("could not pretty print hcl: %w", err)
+	}
+	return string(hBytes), nil
+}
+
+func resourceOrSubresourceHCL(schema map[string]*tfschema.Schema, attributes map[string]string) (string, error) {
```

Comment:
this looks like it's pretty similar to the cnrm way of doing this - in particular by relying on a string builder and hand-crafting each line. I would prefer doing something like https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclwrite#example-package-GenerateFromScratch in the long run, since then we wouldn't have to be responsible for string formatting - but I would potentially be fine with doing it like this for now - I'm just not sure when we would end up upgrading it in the future.

Reply:
turns out hclwrite cannot directly write cty.Value the way we want, and have to do something like this function. Basically, if the value type is an object, we need to create a block, and for others, we can just set them as the block's attributes.

Reply:
I think that's fine as long as we're building a set of Blocks that hclwrite is responsible for rendering (rather than building the hcl string directly)